crates/sui/src/client_commands.rs

```diff
@@ -840,6 +851,47 @@ impl SuiClientCommands {
                 };
                 SuiClientCommandResult::Addresses(output)
             }
+            SuiClientCommands::Balance {
+                address,
+                with_coins,
+            } => {
+                let address = get_identity_address(address, context)?;
+                let client = context.get_client().await?;
+                let mut objects: Vec<Coin> = Vec::new();
+                let mut cursor = None;
+                loop {
+                    let response = client
+                        .coin_read_api()
+                        .get_all_coins(address, cursor, None)
```

Comment:
Initially I was wondering why we were using `get_all_coins` instead of `get_all_balances`, but I see that we do so to support `with_coins = true`. This still feels like a lot of work we are redoing though ... What if instead of `SuiClientCommandResult::Balance(coins_by_type, with_coins)`, the balance result is an enum, 
```
struct BalanceResult {
  Balances(Vec<RpcBalance>),
  BalancesWithCoins(Vec<RpcBalance>, BTreeMap<String, Vec<Coin>>, BTreeMap<String, CoinMetadata>)
}
```

that way you don't have to sum up each coin to calculate the balance - it's already done for you, and you can match on enum variants instead of on the bool

Reply:
Great suggestion, thanks @wlmyng

Reply:
Now I remember why I did not go that route. It returns a balance that has already calculated the balance for all the objects for that coin type and returns the aggregated result. If the `with_coins` is passed, it'd have to query for specific owned objects/coins again. https://docs.sui.io/sui-api-ref#suix_getallbalances

Reply:
What this tells me is that we need to improve our coin read api, probably a better choice is to have 1 api: 
`get_coins(owner: SuiAddress, coin_type: Option<String>, cursor: Option<String>, limit: Option<u64>)` , where the coin_type can be None, which would get all coins. 
The current one gets only `0x2::sui::SUI` when passing None to the call.